src/test/scala/is/hail/TestUtils.scala

```diff
@@ -396,9 +364,48 @@ object TestUtils {
           SafeRow(resultType.asInstanceOf[TBaseStruct], region, resultOff).get(0)
         }
     }
+  }
+
+  def assertEvalsTo(x: IR, expected: Any) {
+    assertEvalsTo(x, Env.empty, FastIndexedSeq(), None, expected)
+  }
+
+  def assertEvalsTo(x: IR, agg: (IndexedSeq[Row], TStruct), expected: Any) {
+    assertEvalsTo(x, Env.empty, FastIndexedSeq(), Some(agg), expected)
+  }
+
+  def assertEvalsTo(x: IR, env: Env[(Any, Type)], args: IndexedSeq[(Any, Type)], agg: Option[(IndexedSeq[Row], TStruct)], expected: Any) {
+    val t = x.typ
+    assert(t.typeCheck(expected))
+
+    val i = Interpret[Any](x, env, args, agg)
+    val i2 = Interpret[Any](x, env, args, agg, optimize = false)
+    val c = eval(x, env, args, agg)
+
+    assert(t.typeCheck(i))
+    assert(t.typeCheck(i2))
+    assert(t.typeCheck(c))
 
-    assert(i == c)
+    assert(t.valuesSimilar(i, expected))
+    assert(t.valuesSimilar(i2, expected))
+    assert(t.valuesSimilar(c, expected))
+  }
+
+  def assertThrows[E <: Throwable : Manifest](x: IR, regex: String) {
+    assertThrows[E](x, Env.empty[(Any, Type)], FastIndexedSeq.empty[(Any, Type)], None, regex)
+  }
+
+  def assertThrows[E <: Throwable : Manifest](x: IR, env: Env[(Any, Type)], args: IndexedSeq[(Any, Type)], agg: Option[(IndexedSeq[Row], TStruct)], regex: String) {
+    interceptException[E](regex)(Interpret[Any](x, env, args, agg))
```

Comment:
are we going to require that the interpreter and the compiler throw the exact same exceptions? :-|

(for the ArrayRange case, stepsize=0 throws a hail exception with a nice error message but the interpreter throws an `IllegalArgumentException` because it's just calling `Range`).

Reply:
I think so, yeah. I've fixed a few of those mismatches in #3554.